sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/intervalExpressions.scala

```diff
@@ -0,0 +1,108 @@
+package org.apache.spark.sql.catalyst.expressions
+
+import java.util.Locale
+
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.catalyst.util.IntervalUtils
+import org.apache.spark.sql.catalyst.util.IntervalUtils._
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.CalendarInterval
+
+abstract class IntervalPart(
+    child: Expression,
+    val dataType: DataType,
+    func: CalendarInterval => Any,
+    funcName: String)
+  extends UnaryExpression with ExpectsInputTypes with Serializable {
+  override def inputTypes: Seq[AbstractDataType] = Seq(CalendarIntervalType)
+  override protected def nullSafeEval(interval: Any): Any = {
+    func(interval.asInstanceOf[CalendarInterval])
+  }
+  override protected def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val iu = IntervalUtils.getClass.getName.stripSuffix("$")
+    defineCodeGen(ctx, ev, c => s"$iu.$funcName($c)")
+  }
+}
+
+case class IntervalMillenniums(child: Expression)
+  extends IntervalPart(child, IntegerType, getMillenniums, "getMillenniums")
+
+case class IntervalCenturies(child: Expression)
+  extends IntervalPart(child, IntegerType, getCenturies, "getCenturies")
+
+case class IntervalDecades(child: Expression)
+  extends IntervalPart(child, IntegerType, getDecades, "getDecades")
+
+case class IntervalYears(child: Expression)
+  extends IntervalPart(child, IntegerType, getYears, "getYears")
+
+case class IntervalQuarters(child: Expression)
+  extends IntervalPart(child, ByteType, getQuarters, "getQuarters")
+
+case class IntervalMonths(child: Expression)
+  extends IntervalPart(child, ByteType, getMonths, "getMonths")
+
+case class IntervalDays(child: Expression)
+  extends IntervalPart(child, LongType, getDays, "getDays")
+
+case class IntervalHours(child: Expression)
+  extends IntervalPart(child, ByteType, getHours, "getHours")
+
+case class IntervalMinutes(child: Expression)
+  extends IntervalPart(child, ByteType, getMinutes, "getMinutes")
+
+case class IntervalSeconds(child: Expression)
+  extends IntervalPart(child, DecimalType(8, 6), getSeconds, "getSeconds")
+
+case class IntervalMilliseconds(child: Expression)
+  extends IntervalPart(child, DecimalType(8, 3), getMilliseconds, "getMilliseconds")
+
+case class IntervalMicroseconds(child: Expression)
+  extends IntervalPart(child, LongType, getMicroseconds, "getMicroseconds")
+
+// Number of seconds in 10000 years is 315576000001 (30 days per one month)
+// which is 12 digits + 6 digits for the fractional part of seconds.
+case class IntervalEpoch(child: Expression)
+  extends IntervalPart(child, DecimalType(18, 6), getEpoch, "getEpoch")
+
+object IntervalPart {
+
+  def parseExtractField(
+      extractField: String,
+      source: Expression,
+      errorHandleFunc: => Nothing): Expression = extractField.toUpperCase(Locale.ROOT) match {
+    case "MILLENNIUM" | "MILLENNIA" | "MIL" | "MILS" => IntervalMillenniums(source)
```

Comment:
shall we put `extract` in the name? e.g. `ExtractIntervalMillenniums`.

Reply:
Usually, packages are used to organize classes to namespaces. I don't like the approach when namespaces are included to class names. Originally I put the interval expression classes to the separate package but had to follow @dongjoon-hyun suggestion https://github.com/apache/spark/pull/25981#discussion_r329869087 . If you think that one more word makes those class names better, I will do that. But I would prefer put the class to the `org.apache.spark.sql.catalyst.expressions.interval` package with the same names as we already have for date-time expressions like: `org.apache.spark.sql.catalyst.expressions.interval.Millennium`, or maybe `org.apache.spark.sql.catalyst.expressions.interval.ExtractMillennium`.

Reply:
I have added the `Extract` prefix to all classes